src/main/scala/com/microsoft/partnercatalyst/fortis/spark/Pipeline.scala

```diff
@@ -0,0 +1,100 @@
+      val locationsExtractorFactory = new LocationsExtractorFactory(featureServiceClient, geofence).buildLookup()
+      val locationFetcher = locationsExtractorFactory.fetch _
+      val keywordExtractor = new KeywordExtractor(List("Ariana"))
+      val imageAnalyzer = new ImageAnalyzer(ImageAnalysisAuth(Settings.oxfordVisionToken), featureServiceClient)
+      val languageDetector = new LanguageDetector(LanguageDetectorAuth(Settings.oxfordLanguageToken))
+      val sentimentDetectorAuth = SentimentDetectorAuth(Settings.oxfordLanguageToken)
+      val supportedLanguages = Set("en", "fr", "de")
+
+      def convertToSchema(original: T): ExtendedFortisEvent[T] = {
+        val message = analyzer.toSchema(original, locationFetcher, imageAnalyzer)
+        ExtendedFortisEvent(message, Analysis())
+      }
+
+      def addLanguage(event: ExtendedFortisEvent[T]): ExtendedFortisEvent[T] = {
+        val language = analyzer.detectLanguage(event.details, languageDetector)
+        event.copy(analysis = Analysis(language = language))
+      }
+
+      def isLanguageSupported(analysis: Analysis): Boolean = {
+        analysis.language match {
+          case None => false
+          case Some(language) => supportedLanguages.contains(language)
+        }
+      }
+
+      def addKeywords(event: ExtendedFortisEvent[T]): ExtendedFortisEvent[T] = {
+        val keywords = analyzer.extractKeywords(event.details, keywordExtractor)
+        event.copy(analysis = event.analysis.copy(keywords = keywords))
+      }
+
+      def hasKeywords(analysis: Analysis): Boolean = {
+        analysis.keywords.nonEmpty
+      }
+
+      def addEntities(event: ExtendedFortisEvent[T]): ExtendedFortisEvent[T] = {
+        val entities = analyzer.extractEntities(event.details, new PeopleRecognizer(modelsProvider, event.analysis.language))
+        event.copy(analysis = event.analysis.copy(entities = entities))
+      }
+
+      def addSentiments(event: ExtendedFortisEvent[T]): ExtendedFortisEvent[T] = {
+        val sentiments = analyzer.detectSentiment(event.details,
+          new SentimentDetector(modelsProvider, event.analysis.language, sentimentDetectorAuth))
+        event.copy(analysis = event.analysis.copy(sentiments = sentiments))
+      }
+
+      def addLocations(event: ExtendedFortisEvent[T]): ExtendedFortisEvent[T] = {
+        val locations = analyzer.extractLocations(event.details,
+          locationsExtractorFactory.create(Some(new PlaceRecognizer(modelsProvider, event.analysis.language))))
+        event.copy(analysis = event.analysis.copy(locations = locations))
+      }
+
+      // Configure analysis pipeline
+      rdd
+        .map(convertToSchema)
+        .map(addLanguage)
+        .filter(item => isLanguageSupported(item.analysis))
+        .map(addKeywords)
+        .filter(item => hasKeywords(item.analysis))
+        .map(item => addLocations(addSentiments(addEntities(item))))
```

Comment:
In what way does the previous code avoid a data dependency? Each stage in the existing code takes an `AnalyzedItem` and outputs a new instance of an `AnalyzedItem`. How can Spark do any of those stages in parallel?

Aside, I'm not convinced that scaling the analysis of a single event across multiple workers is beneficial. Our scaling comes from the ability to consume many events in parallel. Do we really expect to have enough resources to assign N > 1 nodes per incoming event? I think our best bet for optimization of these three calls in general is to use a single `map` call and write our own multi-threaded logic there to execute on a single worker. This implementation can easily accommodate that as well.